### PR TITLE
CLN: algos

### LINKED ISSUE #15903: CLN: fix core/algorithms.py dtype detection
after this is merged: https://github.com/pandas-dev/pandas/pull/15299

we have separate if/else detections for pretty much every routines, e.g. things like

```
if needs_i8_conversion:
    ....
elif .....
```

This is very smelly and needs to be refactored

#15902 instroduces ``_ensure_data`` to do this, but is restricted ATM to nlargest/nsamllest and needs integration. 

Note that this is actually a bit tricky, because we actually want to return a bit more data, including a previously localized tz (for re-localization after).

### PR DESCRIPTION
**Pull request by jreback**
- clean up select_n algos
- clean ensure_data
closes #15903

should make this much simpler going forward. All dtype conversions are now centralized.
Added some doc-strings as well.

### CODE REVIEW

**Comment by jreback** [comment]
hey actually decreased lines even though I added comments, whoo hoo!

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -45,40 +37,192 @@
...
+def _ensure_data(values, dtype=None):
     """
-    helper routine to ensure that our data is of the correct
+    routine to ensure that our data is of the correct
     input dtype for lower-level routines
 
+    This will coerce:
+    - ints -> int64
+    - uint -> uint64
+    - bool -> uint8
```
Comment (chris-b1) [inline pandas/core/algorithms.py]: bool is being coerced to `uint64` below

File: pandas/core/algorithms.py [hunk 2]
```diff
@@ -45,40 +37,192 @@
...
+        elif is_timedelta64_dtype(values) or is_timedelta64_dtype(dtype):
+            from pandas import TimedeltaIndex
+            values = TimedeltaIndex(values)
+            dtype = values.dtype
+        elif is_datetimetz(values) or is_datetime64tz_dtype(dtype):
+            from pandas import DatetimeIndex
+            values = DatetimeIndex(values)
+            dtype = values.dtype
+        else:
+            from pandas import DatetimeIndex
```
Comment (chris-b1) [inline pandas/core/algorithms.py]: Shouldn't this else branch never be hit?

File: pandas/core/algorithms.py [hunk 3]
```diff
@@ -45,40 +37,192 @@
...
+            values = DatetimeIndex(values)
+            dtype = values.dtype
+
+        return values.asi8.view('int64'), dtype, 'int64'
```
Comment (chris-b1) [inline pandas/core/algorithms.py]: `view` is redundant with `asi8`?

File: pandas/core/algorithms.py [hunk 4]
```diff
@@ -45,40 +37,192 @@
...
+            ndtype = dtype = 'float64'
+        elif is_float_dtype(values) or is_float_dtype(dtype):
+            values = _ensure_float64(values)
+            ndtype = dtype = 'float64'
+        else:
+            values = _ensure_object(values)
+            ndtype = dtype = 'object'
+
+    except (TypeError, ValueError):
+        # object array conversion will fail
```
Comment (chris-b1) [inline pandas/core/algorithms.py]: not clear what would actually hit this? should't object fall to the `else` branch above?

File: pandas/core/algorithms.py [hunk 5]
```diff
@@ -513,60 +557,45 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
...
-    elif is_float_dtype(dtype):
-        values = _ensure_float64(values)
-        keys, counts = htable.value_count_float64(values, dropna)
     else:
-        values = _ensure_object(values)
-        keys, counts = htable.value_count_object(values, dropna)
+        # ndarray like
+
+        # TODO: handle uint8
+        f = getattr(htable, "value_count_{dtype}".format(dtype=ndtype))
```
Comment (chris-b1) [inline pandas/core/algorithms.py]: Not a big deal, but seems a little strange that `value_counts_...` and `duplicated_...` are free functions and not based on hashtable classes? Would unify the the dispatch.

File: pandas/core/algorithms.py [hunk 6]
```diff
@@ -45,40 +37,192 @@
...
+
+    Parameters
+    ----------
+    values : ndarray
+    dtype : pandas_dtype
+    original : ndarray-like
 
+    Returns
+    -------
+    Index or ndarray casted to dtype
```
Comment (chris-b1) [inline pandas/core/algorithms.py]: Add that Index is only for extension dtypes only (correct?)  In other words no reboxing if original values were Index.

File: pandas/core/algorithms.py [hunk 7]
```diff
@@ -887,261 +892,186 @@ def _get_score(at):
...
-
-
-def select_n_series(series, n, keep, method):
-    """Implement n largest/smallest for pandas Series
+    Implement n largest/smallest for Series
 
     Parameters
     ----------
-    series : pandas.Series object
+    frame : pandas.DataFrame object
```
Comment (chris-b1) [inline pandas/core/algorithms.py]: Series

**Review by chris-b1 — COMMENTED** [review]
Looks good, significant cleanup!

File: pandas/core/algorithms.py (hunk 1 above)
Comment (jreback) [inline pandas/core/algorithms.py]: yeah that was added subsequently

File: pandas/core/algorithms.py (hunk 5 above)
Comment (jreback) [inline pandas/core/algorithms.py]: not sure what you mean. These are cython functions.

File: pandas/core/algorithms.py (hunk 2 above)
Comment (jreback) [inline pandas/core/algorithms.py]: this is for plain old M8[ns] (IOW regular datetime that is not tz aware).

Comment (jreback) [inline pandas/core/algorithms.py]: but its the same so combined.

File: pandas/core/algorithms.py (hunk 6 above)
Comment (jreback) [inline pandas/core/algorithms.py]: done

File: pandas/core/algorithms.py (hunk 7 above)
Comment (jreback) [inline pandas/core/algorithms.py]: done

File: pandas/core/algorithms.py (hunk 3 above)
Comment (jreback) [inline pandas/core/algorithms.py]: done

File: pandas/core/algorithms.py (hunk 4 above)
Comment (jreback) [inline pandas/core/algorithms.py]:
```
In [1]: np.array([True, False])
Out[1]: array([ True, False], dtype=bool)

In [2]: np.array([True, False]).view('uint64')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-c601d9ef581a> in <module>()
----> 1 np.array([True, False]).view('uint64')

ValueError: new type not compatible with array.
```

though I changed it a different way as ``.astype('uint64')`` works

Comment (jreback) [inline pandas/core/algorithms.py]: and this fall thru is actually for this

```
In [2]: np.array(['a']).astype('int64')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-593402113270> in <module>()
----> 1 np.array(['a']).astype('int64')

ValueError: invalid literal for int() with base 10: 'a'
```

IOW, we are attempting to coerce to a passed dtype, but the input is incompat.

File: pandas/core/algorithms.py [hunk 8]
```diff
@@ -104,92 +246,41 @@ def match(to_match, values, na_sentinel=-1):
...
 
-        uniques = table.unique(_ensure_object(values))
+    table = htable(len(values))
+    uniques = table.unique(values)
+    uniques = _reconstruct_data(uniques, dtype, original)
 
     return uniques
 
 
+unique = unique1d
```
Comment (jorisvandenbossche) [inline pandas/core/algorithms.py]: This is exposed at the top-level, so can you keep the docstring of `unique` ? (put it in `unique1D`)

**Review by jorisvandenbossche — COMMENTED** [review]
Looks like a nice clean-up indeed!

File: pandas/core/algorithms.py (hunk 8 above)
Comment (jreback) [inline pandas/core/algorithms.py]: yes, i'll push it.

File: pandas/core/algorithms.py [hunk 9]
```diff
@@ -45,40 +37,190 @@
...
+    if ndtype == 'object':
+
+        # its cheaper to use a String Hash Table than Object
+        if lib.infer_dtype(values) in ['string']:
+            try:
+                f = func_map['string']
+            except KeyError:
+                pass
+
+    f = func_map.get(ndtype, func_map['object'])
```
Comment (OXPHOS) [inline pandas/core/algorithms.py]: I believe this `f` should be under an `else`:
```
if ndtype == 'object':
    if lib.infer_dtype(values) in ['string']:
       f = func_map['string']
    else:
       f = func_map.get()
```

File: pandas/core/algorithms.py [hunk 10]
```diff
@@ -45,40 +37,190 @@
...
+    if is_categorical_dtype(values):
+        values = values._values_for_rank()
+
+    values, dtype, ndtype = _ensure_data(values)
+    if ndtype == 'object':
+
+        # its cheaper to use a String Hash Table than Object
+        if lib.infer_dtype(values) in ['string']:
+            try:
+                f = func_map['string']
```
Comment (OXPHOS) [inline pandas/core/algorithms.py]: EDIT: The better way seems to be returning `f, values` here

Comment (jreback) [inline pandas/core/algorithms.py]: thanks
https://github.com/pandas-dev/pandas/pull/16128

**Pull request merged**